be/src/exec/pipeline/jdbc_scan_operator.cpp

```diff
@@ -0,0 +1,223 @@
+        return false;
+    }
+    // some chunks are not being consumed
+    if (!_result_chunks.empty()) {
+        return true;
+    }
+
+    // no new chunks and scanner thread finished successfully
+    Status scanner_status;
+    bool scanner_finished = _is_scanner_finished(&scanner_status);
+    if (scanner_finished && scanner_status.ok()) {
+        return false;
+    }
+    // Here are two situation
+    // 1. scanner thread is still running
+    // 2. scanner thread finished in error, we should make sure `pull_chunk` has a chance to get scanner status
+    return true;
+}
+
+bool JDBCScanOperator::pending_finish() const {
+    return !_is_finished.load();
+}
+
+bool JDBCScanOperator::is_finished() const {
+    Status scanner_status;
+    bool scanner_finished = _is_scanner_finished(&scanner_status);
+    if (scanner_finished) {
+        // scanner thread finished and all chunks have been consumed
+        if (scanner_status.ok() && _result_chunks.empty()) {
+            return true;
+        }
+        // scanner thread finished in error, we should make sure `pull_chunk` has a chance to get scanner status
+        if (!scanner_status.ok()) {
+            return false;
+        }
+    }
+    return _is_finished.load();
+}
+
+Status JDBCScanOperator::_start_scanner_thread(RuntimeState* state) {
+    _scanner_thread.reset(new std::thread(&JDBCScanOperator::_start_scanner, this, state));
+    return Status::OK();
+}
+
+void JDBCScanOperator::_start_scanner(RuntimeState* state) {
+    auto tuple_desc = state->desc_tbl().get_tuple_descriptor(_jdbc_scan_node.tuple_id);
+    const auto* jdbc_table = dynamic_cast<const JDBCTableDescriptor*>(tuple_desc->table_desc());
+
+    Status status;
+    vectorized::JDBCScanContext scan_ctx;
+    std::string driver_name = jdbc_table->jdbc_driver_name();
+    std::string driver_url = jdbc_table->jdbc_driver_url();
+    std::string driver_checksum = jdbc_table->jdbc_driver_checksum();
+    std::string driver_class = jdbc_table->jdbc_driver_class();
+    std::string driver_location;
+
+    if (status = JDBCDriverManager::getInstance()->get_driver_location(driver_name, driver_url, driver_checksum, &driver_location); !status.ok()) {
+        LOG(ERROR) << fmt::format("Get JDBC Driver[{}] error, error is {}", driver_name, status.to_string());
+        _set_scanner_state(true, status);
+        return;
+    }
+
+    scan_ctx.driver_path = driver_location;
+    scan_ctx.driver_class_name = driver_class;
+    scan_ctx.jdbc_url = jdbc_table->jdbc_url();
+    scan_ctx.user = jdbc_table->jdbc_user();
+    scan_ctx.passwd = jdbc_table->jdbc_passwd();
+    scan_ctx.sql = get_jdbc_sql(jdbc_table->jdbc_table(), _jdbc_scan_node.columns, _jdbc_scan_node.filters, _limit);
+
+    _scanner.reset(new vectorized::JDBCScanner(scan_ctx, _result_tuple_desc));
+
+    if (status = _scanner->open(state); !status.ok()) {
+        _set_scanner_state(true, status);
+        _scanner->close(state);
+        return;
```

Comment:
Better to return `Status` and use `RETURN_IF_ERROR` at upper layer

Reply:
> Better to return `Status` and use `RETURN_IF_ERROR` at upper layer

because the entire `_start_scanner` function is executed by the async _scanner_thread, `JDBCScanOperator` obtains the execution result through `_is_scanner_finished` interface, we have no chance and no need to handle the return value, so I think it is ok to return void